syntax = "proto3";

package pujpb;

enum FuzzyRule {
  FR_NONE = 0;
  FR_UR_As_U = 1;
  FR_OR_As_O = 2;
  FR_OR_As_E = 3;
  FR_ORH_As_OH = 4;
  FR_EU_As_IU = 6;
  FR_OINN_As_AINN = 7;
  FR_UOINN_As_UINN = 8;
  FR_UOINN_As_UAINN = 9;
  FR_OI_As_UE = 10;
  FR_OU_As_AU = 11;
  FR_UE_As_UEI = 12;
  FR_URN_As_IN = 13;
  FR_IN_As_EN = 14;
  FR_UENG_As_ENG_UANG = 15;
  FR_UEK_As_UAK = 16 [deprecated = true];
  FR_IO_As_IE = 17;
  FR_IAU_As_IEU = 18;
  FR_IAU_As_IOU = 19;
  FR_IAN_As_IEN = 20;
  FR_UAN_As_UEN = 21;
  FR_IAM_As_IEM = 22;
  FR_N_As_L_ForMEnding = 23;
  FR_N_As_L_ForNOrNGEnding = 24 [deprecated = true];
  FR_L_As_N_ForMEnding = 25;
  FR_MU_As_BU_ForNasalEnding = 26;
  FR_BU_As_MU_ForNasalEnding = 27;
  FR_N_As_NG = 28;
  FR_M_As_NG = 29;
  FR_ENG_As_EN = 30;
  FR_IM_As_ING = 31;
  FR_IN_As_ING = 32;
  FR_URN_As_URNG = 33;
  FR_UN_As_UNG = 34;
  FR_AM_As_ANG = 35;
  FR_AN_As_ANG = 36;
  FR_IAM_As_IANG = 37;
  FR_IAN_As_IANG = 38;
  FR_UAM_As_UANG = 39;
  FR_UAN_As_UANG = 40;
  FR_NG_As_UNG = 41;
  FR_NG_As_URNG = 42;
  FR_IONG_As_ONG = 43;
  FR_NGU_As_U = 44;
}

enum ToneSpecial {
  TS_NONE = 0;
  TS_SMOOTH_2ND_3RD_4TH = 1;
  TS_SMOOTH_NEUTRAL = 2;
  TS_VARIABLE_3RD_2ND = 3;
}

message FuzzyRuleAction {
  // 目前总共两类：initial；initial+final
  string action = 1;
  // 正则表达式匹配串
  string pattern = 2;
  // 替换表达式，包含 $1 $2 $3 用于 Javascript.
  string replacement_dollar = 3;
  // 替换表达式，包含 \1 \2 \3 用于 Python.
  string replacement_backslash = 4;
}

message FuzzyRuleDescriptor {
  FuzzyRule id = 1;
  repeated FuzzyRuleAction actions = 2;
}

message Tones {
  repeated uint32 citation = 1;
  repeated uint32 sandhi = 2;
  repeated uint32 neutral = 3;
  repeated ToneSpecial specials = 4;
}

message Accent {
  string id = 1;
  string area = 2;
  optional string subarea = 3;
  repeated FuzzyRule rules = 4;
  reserved 5;
  Tones tones = 6;
}

message Accents {
  repeated Accent accents = 1;
  repeated FuzzyRuleDescriptor fuzzy_rule_descriptors = 2;
}
